src/jira/util/jwt.ts

```diff
@@ -94,25 +131,16 @@ export const hasValidJwt = (secret: string, baseUrl: string, req: Request, res:
 		return false;
 	}
 
-	// First check query string params
 	if (verifiedClaims.qsh) {
-		let expectedHash = jwt.createQueryStringHash(req, false, baseUrl);
-		let signatureHashVerified = verifiedClaims.qsh === expectedHash;
-		if (!signatureHashVerified) {
-			jwt.createCanonicalRequest(req, false, baseUrl); // eslint-disable-line
-
-			// If that didn't verify, it might be a post/put - check the request body too
-			expectedHash = jwt.createQueryStringHash(req, true, baseUrl);
-			signatureHashVerified = verifiedClaims.qsh === expectedHash;
-			if (!signatureHashVerified) {
-				jwt.createCanonicalRequest(req, true, baseUrl); // eslint-disable-line
-
-				// Send the error message for the first verification - it's 90% more likely to be the one we want.
-				sendError(res, 401, "Authentication failed: query hash does not match.");
-				return false;
-			}
+		if (tokenType === TokenType.context) {
```

Comment:
So these changes we can feature flag:
1. We always require qsh to be set
2. If token type is "context" we expect it to be equal to "context-qsh"

Reply:
[] add a feature flag when Tom's changes are merged

Reply:
Do we really need the token type?  Couldn't you just do `return verifiedClaims.qsh === "context-qsh" || verifyQsh(verifiedClaims.qsh, baseUrl, req, res);`?  Why do we need to specify it?

Reply:
Yes, we do. For the security reason we need to accept either one or another way of authentication. 

It is done so we'll have only dedicated endpoints accessible with the context token "context-qsh". If somebody steal token with "context-qsh" qsh, then this person can get access to any endpoint if we don't differentiate. Otherwise other endpoints are protected by requiring strict qsh.

There are more details about it here: https://community.developer.atlassian.com/t/action-required-atlassian-connect-vulnerability-allows-bypass-of-app-qsh-verification-via-context-jwts/47072